consensus/qsc/broadcast.go

```diff
@@ -0,0 +1,341 @@
+	mino.UnsupportedHandler
+
+	ch    chan *MessageSet
+	store *storage
+}
+
+// Process implements mino.Handler. It handles two cases: (1) A message set sent
+// from a player that must be processed. (2) A message set request that returns
+// the list of messages missing to the distant player.
+func (h hTLCR) Process(in proto.Message) (proto.Message, error) {
+	switch msg := in.(type) {
+	case *MessageSet:
+		h.ch <- msg
+	case *RequestMessageSet:
+		if msg.GetTimeStep() != h.store.previous.GetTimeStep() {
+			return nil, nil
+		}
+
+		resp := proto.Clone(h.store.previous).(*MessageSet)
+		for _, node := range msg.GetNodes() {
+			// Remove those not necessary to reduce the message size.
+			delete(resp.Messages, node)
+		}
+
+		return h.store.previous, nil
+	}
+
+	return nil, nil
+}
+
+type tlcr interface {
+	execute(context.Context, ...*Message) (*View, error)
+}
+
+// bTLCR implements the TLCR primitive from the QSC paper.
+type bTLCR struct {
+	logger   zerolog.Logger
+	node     int64
+	timeStep uint64
+	players  mino.Players
+	rpc      mino.RPC
+	store    *storage
+	// Mino handler impl should redirect the messages to this channel..
+	ch chan *MessageSet
+}
+
+func newTLCR(name string, node int64, mino mino.Mino, players mino.Players) (*bTLCR, error) {
+	// TODO: improve to have a buffer per node with limited size.
+	ch := make(chan *MessageSet, 1000)
+	store := &storage{}
+	rpc, err := mino.MakeRPC(name, hTLCR{ch: ch, store: store})
+	if err != nil {
+		return nil, err
+	}
+
+	tlcr := &bTLCR{
+		logger:   fabric.Logger,
+		node:     node,
+		timeStep: 0,
+		rpc:      rpc,
+		ch:       ch,
+		players:  players,
+		store:    store,
+	}
+
+	return tlcr, nil
+}
+
+func (b *bTLCR) execute(ctx context.Context, messages ...*Message) (*View, error) {
+	ms := &MessageSet{
+		Messages: make(map[int64]*Message),
+		TimeStep: b.timeStep,
+		Node:     b.node,
+	}
+	for _, msg := range messages {
+		ms.Messages[msg.GetNode()] = msg
+	}
+
+	_, errs := b.rpc.Call(ms, b.players)
+
+	for len(ms.GetMessages()) < b.players.Len() {
+		select {
+		case err := <-errs:
+			b.logger.Err(err).Msg("couldn't broadcast to everyone")
+		case req := <-b.ch:
+			if req.GetTimeStep() == b.timeStep {
+				b.merge(ms, req)
+			} else {
+				err := b.catchUp(ms, req)
+				if err != nil {
+					b.logger.Err(err).Send()
+				}
+			}
+		case <-ctx.Done():
+			return nil, xerrors.Errorf("context is done: %v", ctx.Err())
+		}
+	}
+
+	b.timeStep++
+	b.logger.Trace().Msgf("node %d moving to time step %d", b.node, b.timeStep)
+
+	b.store.previous = ms
+
+	return &View{Received: ms.GetMessages()}, nil
+}
+
+// merge merges received into curr.
+func (b *bTLCR) merge(current, received *MessageSet) {
+	for k, v := range received.GetMessages() {
+		current.Messages[k] = v
+	}
+}
+
+// catchUp analyses the received message set compared to the current one and
+// tries to catch up if possible, otherwise it delays the processing by
+// inserting the message at the end of the queue.
```

Comment:
```suggestion
// catchUp analyses the messageSet 'received' compared to 'current' and tries to
// catch up if possible, otherwise it delays the processing by inserting the
// message at the end of the queue.
```